docs/configuration.md
# Configuration Guide

## Summary

This SDK uses a structure called "Config" to store and manage configuration, read ["config/config.go"](https://github.com/hewenxiang/shanhe-sdk-go/blob/master/config/config.go) file's comments of public functions for more information.

Except for AccessKeyID and SecretAccessKey, you can also configure the API servers for private cloud usage scenario. All available configurable items are list in default configuration file.

___Default Configuration File:___

``` yaml
# QingCloud services configuration

qy_access_key_id: 'ACCESS_KEY_ID'
qy_secret_access_key: 'SECRET_ACCESS_KEY'

host: 'api.qingcloud.com'
port: 443
protocol: 'https'
uri: '/iaas'
connection_retries: 3

# Valid log levels are "debug", "info", "warn", "error", and "fatal".
log_level: 'warn'

```

## Usage

1. Just create a config structure instance with your API AccessKey, and initialize services that you need using Init() function of target service.

2. Or if you do not want to expose your AccessKeyID and SecretAccessKey, you can also call our API without them in the following way:
- Go to our IAM service, create an instance role and attach it to your instance.
- Create the config structure instance without AccessKeyID and SecretAccessKey.

Note that you can also configure your own credential proxy server (where you retrieve tokens) in configuration file like this:

```yaml
credential_proxy_protocol: 'http'
credential_proxy_host: '169.254.169.254'
credential_proxy_port: 80
credential_proxy_uri: '/latest/meta-data/security-credentials'
```

### Code Snippet

Create default configuration

``` go
defaultConfig, _ := config.NewDefault()
```

Create configuration from AccessKey

``` go
configuration, _ := config.New("ACCESS_KEY_ID", "SECRET_ACCESS_KEY")

anotherConfiguration := config.NewDefault()
anotherConfiguration.AccessKeyID = "ACCESS_KEY_ID"
anotherConfiguration.SecretAccessKey = "SECRET_ACCESS_KEY"
```

Load user configuration

``` go
userConfig, _ := config.NewDefault().LoadUserConfig()
```

Load configuration from config file

``` go
configFromFile, _ := config.NewDefault().LoadConfigFromFilepath("PATH/TO/FILE")
```

Change API server

``` go
moreConfiguration, _ := config.NewDefault()

moreConfiguration.Protocol = "https",
moreConfiguration.Host = "api.private.com",
moreConfiguration.Port = 4433,
moreConfiguration.URI = "/iaas",
```


CHANGELOG.md
# Change Log
All notable changes to QingCloud SDK for Go will be documented in this file.

## [v2.0.0-alpha.29] - 2018-03-26

### Added

- CeaseInstances @calvinyv

## [v2.0.0-alpha.28] - 2018-01-08

### Fixed

- DescribeLoadBalancerBackends field error  @runzexia

## [v2.0.0-alpha.27] - 2018-01-08

### Fixed

- ModifyLoadBalancerBackendAttributes field error  @runzexia

## [v2.0.0-alpha.26] - 2018-01-04

### Fixed

- fix loadblancer listener timeout location  

## [v2.0.0-alpha.25] - 2018-01-04

### Fixed

- add loadblancer listener timeout  

## [v2.0.0-alpha.24] - 2018-01-04

### Fixed

- add ModifyLoadBalancerListenerAttributes missing field listener_option

## [v2.0.0-alpha.23] - 2017-12-24

### Fixed

- DescribeServerCertificatesInput field type error

## [v2.0.0-alpha.22] - 2017-12-23

### Add

- Add api to get vpn certificate

### Fixed

- Add loadbalancer attr cluster 
- POST query escape
- CreateServerCertificate use POST

## [v2.0.0-alpha.21] - 2017-12-15

### Fixed

- Change loadbalancer attr cluster to eips


## [v2.0.0-alpha.20] - 2017-12-15

### Fixed

- Fix loadbalancer http_header_size location


## [v2.0.0-alpha.19] - 2017-12-15

### Fixed

- Add loadbalancer http_header_size

## [v2.0.0-alpha.18] - 2017-12-14

### Fixed

- Add loadbalancer node_count


## [v2.0.0-alpha.17] - 2017-12-13

### Fixed

- Fix new config with endpoint

## [v2.0.0-alpha.16] - 2017-12-12

### Added

- Add new config with endpoint

### Fixed

- Fix cluster api bugs
- Fix userdata_path. userdata_file default value error 

## [v2.0.0-alpha.15] - 2017-12-07

### Added

- Add service cluster for go sdk

### Fixed

- Add requset type POST support 

## [v2.0.0-alpha.14] - 2017-11-24

### Fixed

- Add router static missing field
- Update template to fit snips change

## [v2.0.0-alpha.13] - 2017-11-09

### Fixed

- Delete DescribeSecurityGroupRules SecurityGroup Required

## [v2.0.0-alpha.12] - 2017-11-09

### Fixed

- DescribeSecurityGroupRules direction default value

## [v2.0.0-alpha.11] - 2017-11-08

### Fixed

- Fix wrong type of job id for stop lb output
- Add Makefile generate target to generate service code
- Fix ModifySecurityGroupRuleAttributes field type error
- Fix SecurityGroupRule field missing error

## [v2.0.0-alpha.10] - 2017-08-28

### Fixed

- Fixed loadbalancers section in request of StopLoadBalancers

## [v2.0.0-alpha.9] - 2017-08-23

### Added

- Add vxnetid and loadbalancertype parms for load balancer

### Fixed

- Fixed vxnet section in response of DescribeInstances

## [v2.0.0-alpha.8] - 2017-08-13

### Added

- Add missing parameter for describenic
- Add vxnet parms for instances

## [v2.0.0-alpha.7] - 2017-08-02

### Added

- Add timeout parameter for http client
- Add missing parameters in nic, router and security groups

## [v2.0.0-alpha.6] - 2017-07-17

### Added

- Update advanced client. [@jolestar]
- Fix several APIs. [@jolestar]

## [v2.0.0-alpha.5] - 2017-03-27

### Added

- Add advanced client for simple instance management. [@jolestar]
- Add wait utils for waiting a job to finish. [@jolestar]

## [v2.0.0-alpha.4] - 2017-03-14

### Fixed

- Fix Features type in RouterVxNet.

## [v2.0.0-alpha.3] - 2017-01-15

### Changed

- Fix request signer.

## [v2.0.0-alpha.2] - 2017-01-05

### Changed

- Fix logger output format, don't parse special characters.
- Rename package "errs" to "errors".

### Added

- Add type converters.

### BREAKING CHANGES

- Change value type in input and output to pointer.

## v2.0.0-alpha.1 - 2016-12-03

### Added

- QingCloud SDK for the Go programming language.
[v2.0.0-alpha.29]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.28...v2.0.0-alpha.29
[v2.0.0-alpha.28]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.27...v2.0.0-alpha.28
[v2.0.0-alpha.27]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.26...v2.0.0-alpha.27
[v2.0.0-alpha.26]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.25...v2.0.0-alpha.26   
[v2.0.0-alpha.25]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.24...v2.0.0-alpha.25  
[v2.0.0-alpha.24]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.23...v2.0.0-alpha.24  
[v2.0.0-alpha.23]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.22...v2.0.0-alpha.23    
[v2.0.0-alpha.22]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.21...v2.0.0-alpha.22    
[v2.0.0-alpha.21]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.20...v2.0.0-alpha.21  
[v2.0.0-alpha.20]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.19...v2.0.0-alpha.20  
[v2.0.0-alpha.19]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.18...v2.0.0-alpha.19  
[v2.0.0-alpha.18]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.17...v2.0.0-alpha.18  
[v2.0.0-alpha.17]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.16...v2.0.0-alpha.17  
[v2.0.0-alpha.16]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.15...v2.0.0-alpha.16  
[v2.0.0-alpha.15]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.14...v2.0.0-alpha.15  
[v2.0.0-alpha.14]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.13...v2.0.0-alpha.14  
[v2.0.0-alpha.13]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.12...v2.0.0-alpha.13  
[v2.0.0-alpha.12]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.11...v2.0.0-alpha.12  
[v2.0.0-alpha.11]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.10...v2.0.0-alpha.11  
[v2.0.0-alpha.10]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.9...v2.0.0-alpha.10  
[v2.0.0-alpha.9]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.8...v2.0.0-alpha.9  
[v2.0.0-alpha.8]: https://github.com/hewenxiang/shanhe-sdk-go/compare/v2.0.0-alpha.7...v2.0.0-alpha.8  

config/config.go
// +-------------------------------------------------------------------------
// | Copyright (C) 2016 Yunify, Inc.
// +-------------------------------------------------------------------------
// | Licensed under the Apache License, Version 2.0 (the "License");
// | you may not use this work except in compliance with the License.
// | You may obtain a copy of the License in the LICENSE file, or at:
// |
// | http://www.apache.org/licenses/LICENSE-2.0
// |
// | Unless required by applicable law or agreed to in writing, software
// | distributed under the License is distributed on an "AS IS" BASIS,
// | WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// | See the License for the specific language governing permissions and
// | limitations under the License.
// +-------------------------------------------------------------------------

package config

import (
	"fmt"
	"io/ioutil"
	"net"
	"net/http"
	"net/url"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/hewenxiang/shanhe-sdk-go/logger"
	"github.com/hewenxiang/shanhe-sdk-go/utils"
)

// A Config stores a configuration of this sdk.
type Config struct {
	AccessKeyID     string `yaml:"qy_access_key_id"`
	SecretAccessKey string `yaml:"qy_secret_access_key"`

	Host              string `yaml:"host"`
	Port              int    `yaml:"port"`
	Protocol          string `yaml:"protocol"`
	URI               string `yaml:"uri"`
	ConnectionRetries int    `yaml:"connection_retries"`
	ConnectionTimeout int    `yaml:"connection_timeout"`

	LogLevel string `yaml:"log_level"`

	Zone string `yaml:"zone"`


	CredentialProxyProtocol string `yaml:"credential_proxy_protocol"`
	CredentialProxyHost     string `yaml:"credential_proxy_host"`
	CredentialProxyPort     int    `yaml:"credential_proxy_port"`
	CredentialProxyURI      string `yaml:"credential_proxy_uri"`

	Token      string
	Expiration int64

	Connection *http.Client
}

// New create a Config with given AccessKeyID and SecretAccessKey.
func New(accessKeyID, secretAccessKey string) (*Config, error) {
	config, err := NewDefault()
	if err != nil {
		return nil, err
	}

	config.AccessKeyID = accessKeyID
	config.SecretAccessKey = secretAccessKey

	config.Connection = &http.Client{}

	return config, nil
}

// NewWithEndpoint create a Config with given AccessKeyID, SecretAccessKey and endpoint
func NewWithEndpoint(accessKeyID, secretAccessKey, endpoint string) (*Config, error) {
	qcURL, err := url.Parse(endpoint)
	if err != nil {
		return nil, err
	}
	if qcURL.Opaque != "" {
		return nil, fmt.Errorf("wrong URL format")
	}

	if !strings.Contains(qcURL.Host, ":") {
		if qcURL.Scheme == "https" {
			qcURL.Host += ":443"
		} else if qcURL.Scheme == "http" {
			qcURL.Host += ":80"
		} else {
			return nil, fmt.Errorf("can not find default port ")
		}
	}

	config, err := NewDefault()
	if err != nil {
		return nil, err
	}

	host, port, err := net.SplitHostPort(qcURL.Host)
	if err != nil {
		return nil, err
	}

	config.Port, _ = strconv.Atoi(port)
	config.Host = host
	config.AccessKeyID = accessKeyID
	config.SecretAccessKey = secretAccessKey
	config.Protocol = qcURL.Scheme
	config.URI = qcURL.Path
	config.Connection = &http.Client{}
	return config, nil
}

// NewDefault create a Config with default configuration.
func NewDefault() (*Config, error) {
	config := &Config{}
	err := config.LoadDefaultConfig()
	if err != nil {
		return nil, err
	}

	timeout := time.Duration(config.ConnectionTimeout) * time.Second
	transport := &http.Transport{
		Dial: func(network, addr string) (net.Conn, error) {
			return net.DialTimeout(network, addr, timeout)
		},
	}
	config.Connection = &http.Client{
		Transport: transport,
	}

	return config, nil
}

// LoadDefaultConfig loads the default configuration for Config.
// It returns error if yaml decode failed.
func (c *Config) LoadDefaultConfig() error {
	_, err := utils.YAMLDecode([]byte(DefaultConfigFileContent), c)
	if err != nil {
		logger.Error("Config parse error: " + err.Error())
		return err
	}

	logger.SetLevel(c.LogLevel)

	return nil
}

// LoadUserConfig loads user configuration in ~/.qingcloud/config.yaml for Config.
// It returns error if file not found.
func (c *Config) LoadUserConfig() error {
	_, err := os.Stat(GetUserConfigFilePath())
	if err != nil {
		logger.Warn("Installing default config file to \"" + GetUserConfigFilePath() + "\"")
		InstallDefaultUserConfig()
	}

	return c.LoadConfigFromFilepath(GetUserConfigFilePath())
}

// LoadConfigFromFilepath loads configuration from a specified local path.
// It returns error if file not found or yaml decode failed.
func (c *Config) LoadConfigFromFilepath(filepath string) error {
	if strings.Index(filepath, "~/") == 0 {
		filepath = strings.Replace(filepath, "~/", getHome()+"/", 1)
	}

	configYAML, err := ioutil.ReadFile(filepath)
	if err != nil {
		logger.Error("File not found: " + filepath)
		return err
	}

	return c.LoadConfigFromContent(configYAML)
}

// LoadConfigFromContent loads configuration from a given byte slice.
// It returns error if yaml decode failed.
func (c *Config) LoadConfigFromContent(content []byte) error {
	c.LoadDefaultConfig()

	_, err := utils.YAMLDecode(content, c)
	if err != nil {
		logger.Error("Config parse error: " + err.Error())
		return err
	}

	logger.SetLevel(c.LogLevel)

	timeout := time.Duration(c.ConnectionTimeout) * time.Second
	transport := &http.Transport{
		Dial: func(network, addr string) (net.Conn, error) {
			return net.DialTimeout(network, addr, timeout)
		},
	}
	c.Connection = &http.Client{
		Transport: transport,
	}

	return nil
}
